src/transformStateWithClones.js

```diff
@@ -7,7 +7,45 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let addProperties = {};
+  let removeProperties = {};
+  let cleanProperties = {};
+  let addObject = state;
+
+  const array = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      addProperties = { ...addObject, ...action.extraData };
+      array.push(addProperties);
+      addObject = addProperties;
+      continue;
+    }
+
+    if (action.type === 'removeProperties') {
+      removeProperties = { ...addObject };
+
+      for (const i of action.keysToRemove) {
+        delete removeProperties[i];
+      }
+      array.push(removeProperties);
+      addObject = removeProperties;
+      continue;
+    }
+
+    if (action.type === 'clear') {
+      cleanProperties = { ...state };
+
+      for (const keys in cleanProperties) {
+        delete cleanProperties[keys];
+      }
+      array.push(cleanProperties);
+      addObject = cleanProperties;
```

Comment:
The 'clear' action is intended to create an empty state object. There's no need to copy the existing state before clearing it. You can directly assign an empty object to 'cleanProperties'.